File: src/main/java/toby/command/commands/music/SetVolumeCommand.java
package toby.command.commands.music;


import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.InteractionHook;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import toby.command.CommandContext;
import toby.command.ICommand;
import toby.emote.Emotes;
import toby.jpa.dto.UserDto;
import toby.lavaplayer.GuildMusicManager;
import toby.lavaplayer.PlayerManager;

import java.util.List;
import java.util.Optional;

import static toby.command.ICommand.getConsumer;


public class SetVolumeCommand implements IMusicCommand {

    private final String VOLUME = "volume";


    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        handleMusicCommand(ctx, PlayerManager.getInstance(), requestingUserDto, deleteDelay);
    }

    @Override
    public void handleMusicCommand(CommandContext ctx, PlayerManager instance, UserDto requestingUserDto, Integer deleteDelay) {
        SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        ICommand.deleteAfter(event.getHook(), deleteDelay);

        if (!requestingUserDto.hasMusicPermission()) {
            sendErrorMessage(event, deleteDelay);
            return;
        }

        if (IMusicCommand.isInvalidChannelStateForCommand(ctx, deleteDelay)) return;

        final Member member = ctx.getMember();
        setNewVolume(event, instance, member, requestingUserDto, deleteDelay);
    }

    private void setNewVolume(SlashCommandInteractionEvent event, PlayerManager instance, Member member, UserDto requestingUserDto, Integer deleteDelay) {
        int volumeArg = Optional.ofNullable(event.getOption(VOLUME)).map(OptionMapping::getAsInt).orElse(0);
        InteractionHook hook = event.getHook();
        GuildMusicManager musicManager = instance.getMusicManager(event.getGuild());
        if (volumeArg > 0) {
            if (instance.isCurrentlyStoppable() || requestingUserDto.isSuperUser()) {
                AudioPlayer audioPlayer = musicManager.getAudioPlayer();
                if (volumeArg > 100) {
                    hook.sendMessage(getDescription()).setEphemeral(true).queue(getConsumer(deleteDelay));
                    return;
                }
                int oldVolume = audioPlayer.getVolume();
                if (volumeArg == oldVolume) {
                    hook.sendMessageFormat("New volume and old volume are the same value, somebody shoot %s", member.getEffectiveName()).setEphemeral(true).queue(getConsumer(deleteDelay));
                    return;
                }
                audioPlayer.setVolume(volumeArg);
                hook.sendMessageFormat("Changing volume from '%s' to '%s' \uD83D\uDD0A", oldVolume, volumeArg).setEphemeral(true).queue(getConsumer(deleteDelay));
            } else {
                sendErrorMessage(event, deleteDelay);
            }
        } else hook.sendMessage(getDescription()).setEphemeral(true).queue(getConsumer(deleteDelay));
    }

    @Override
    public void sendErrorMessage(SlashCommandInteractionEvent event, Integer deleteDelay) {
        event.getHook().sendMessageFormat("You aren't allowed to change the volume kid %s", event.getGuild().getJDA().getEmojiById(Emotes.TOBY)).setEphemeral(true).queue(getConsumer(deleteDelay));
    }


    @Override
    public String getName() {
        return "setvolume";
    }

    @Override
    public String getDescription() {
        return "Set the volume of the audio player for the server to a percent value (between 1 and 100)";

    }

    @Override
    public List<OptionData> getOptionData() {
        return List.of(new OptionData(OptionType.INTEGER, VOLUME, "Volume value between 1-100 to set the audio to", true));
    }
}

File: discord-bot/src/main/kotlin/bot/toby/command/commands/music/player/SetVolumeCommand.kt
package bot.toby.command.commands.music.player

import bot.toby.command.commands.music.MusicCommand
import bot.toby.emote.Emotes
import bot.toby.lavaplayer.PlayerManager
import core.command.Command.Companion.replyEphemeralAndDelete
import core.command.CommandContext
import database.dto.user.UserDto
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData
import org.springframework.stereotype.Component
import java.util.*

@Component
class SetVolumeCommand : MusicCommand {
    companion object {
        private const val VOLUME = "volume"
    }
    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int) {
        handleMusicCommand(ctx, PlayerManager.instance, requestingUserDto, deleteDelay)
    }

    override fun handleMusicCommand(
        ctx: CommandContext,
        instance: PlayerManager,
        requestingUserDto: UserDto,
        deleteDelay: Int
    ) {
        if (!checkMusicPreconditions(ctx, requestingUserDto, deleteDelay)) return
        setNewVolume(ctx.event, instance, ctx.member, requestingUserDto, deleteDelay)
    }

    private fun setNewVolume(
        event: SlashCommandInteractionEvent,
        instance: PlayerManager,
        member: Member?,
        requestingUserDto: UserDto?,
        deleteDelay: Int
    ) {
        val volumeArg = Optional.ofNullable(event.getOption(VOLUME)).map { obj: OptionMapping -> obj.asInt }.orElse(0)
        val hook = event.hook
        val musicManager = instance.getMusicManager(event.guild!!)
        if (volumeArg > 0) {
            if (instance.isCurrentlyStoppable || requestingUserDto!!.superUser) {
                val audioPlayer = musicManager.audioPlayer
                if (volumeArg > 100) {
                    hook.replyEphemeralAndDelete(description, deleteDelay)
                    return
                }
                val oldVolume = audioPlayer.volume
                if (volumeArg == oldVolume) {
                    hook.replyEphemeralAndDelete(
                        "New volume and old volume are the same value, somebody shoot ${member!!.effectiveName}",
                        deleteDelay,
                    )
                    return
                }
                instance.setPreviousVolume(oldVolume)
                audioPlayer.volume = volumeArg
                hook.replyEphemeralAndDelete(
                    "Changing volume from '$oldVolume' to '$volumeArg' \uD83D\uDD0A",
                    deleteDelay,
                )
            } else {
                sendErrorMessage(event, deleteDelay)
            }
        } else hook.replyEphemeralAndDelete(description, deleteDelay)
    }

    override fun sendErrorMessage(event: SlashCommandInteractionEvent, deleteDelay: Int) {
        event.hook.replyEphemeralAndDelete(
            "You aren't allowed to change the volume kid ${event.guild!!.jda.getEmojiById(Emotes.TOBY)}",
            deleteDelay,
        )
    }

    override val name: String
        get() = "setvolume"
    override val description: String
        get() = "Set the volume of the audio player for the server to a percent value (between 1 and 100)"
    override val optionData: List<OptionData>
        get() = listOf(OptionData(OptionType.INTEGER, VOLUME, "Volume value between 1-100 to set the audio to", true))
}
